src/checkpoint/traits/user_traits.h

```diff
@@ -0,0 +1,196 @@
+#endif
+        std::is_same<std::decay_t<U>,std::decay_t<V>>
+      > {};
+  template<typename U, typename V>
+  inline constexpr bool traits_match_v = traits_match<U,V>::value;
+
+
+
+  template<typename Trait = NoTrait, typename... Traits>
+  struct TraitHolder {
+  protected:
+    template<typename, typename...>
+    friend struct TraitHolder;
+
+    template<typename T>
+    using has_trait = std::disjunction<traits_match<T, Trait>, traits_match<T, Traits>...>;
+
+  public:
+
+    //Remove one at a time, so we can allow multiple trait copies to be removed independently.
+    template<typename T>
+    using _without_trait = std::conditional_t<
+                            traits_match_v<T, Trait>,
+                            TraitHolder<Traits...>,
+                            typename TraitHolder<Traits...>::template _without_trait<T>::template with_pre<Trait>
+                          >;
+    template<typename... T>
+    using with = TraitHolder<Trait, Traits..., T...>;
+    //To respect ordering. Could be handy in the future -- disambiguating multiple hooks?
+    template<typename... T>
+    using with_pre = TraitHolder<T..., Trait, Traits...>;
+
+    //Flush out any nested traits (TraitHolder<..., TraitHolder<...>, ...>)
+    //Remove any NoTraits, unless that's the only trait.
+    using BaseTraits = typename TraitHolder<Traits...>::BaseTraits::template with_pre<Trait>;
+
+    template<typename... T>
+    using has = std::conjunction<has_trait<T>...>;
+    template<typename... T>
+    using has_any = std::disjunction<has_trait<T>...>;
+
+    template<typename... T>
+    static inline constexpr bool has_v = has<T...>::value;
+    template<typename... T>
+    static inline constexpr bool has_any_v = has_any<T...>::value;
+
+    template<typename T, typename... U>
+    using without = typename without_helper<BaseTraits, T, U...>::type;
+  };
+
+  //Empty specialization w/ shortcuts.
+  template<>
+  struct TraitHolder<NoTrait> {
+    using BaseTraits = TraitHolder<NoTrait>;
+
+    template<typename... T>
+    using has = std::false_type;
+    template<typename... T>
+    using has_any = std::false_type;
+
+    template<typename... T>
+    static constexpr bool has_v = false;
+    template<typename... T>
+    static constexpr bool has_any_v = false;
+
+    template<typename... T>
+    using with = TraitHolder<T...>;
+    template<typename... T>
+    using with_pre = TraitHolder<T...>;
+
+    template<typename... T>
+    using without = TraitHolder<NoTrait>;
+
+    template<typename T>
+    using _without_trait = TraitHolder<NoTrait>;
+
+  protected:
+    template<typename T>
+    using has_trait = std::false_type;
+
+    template<typename, typename...>
+    friend struct TraitHolder;
+    template<typename...>
+    friend struct without_helper;
+  };
+
+  //Specialization to remove nested traits.
+  template<typename... Traits, typename... NestedTraits>
+  struct TraitHolder<TraitHolder<NestedTraits...>, Traits...>
```

Comment:
How do these pop up? This feels like it is patching out something weird in a template instantiation somewhere -- do we need to support in the API nested traits that can be collapsed out?

Reply:
I don't believe this does pop up naturally, I just added it on during my first pass thinking it would come in handy. I'll go ahead and drop these

Reply:
Actually, this was helpful in the VT resilience updates to make a new checkpoint::serialize call with the same traits as another serializer